core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala

```diff
@@ -414,9 +425,48 @@ private[spark] class TaskSchedulerImpl(
             launchedAnyTask |= launchedTaskAtCurrentMaxLocality
           } while (launchedTaskAtCurrentMaxLocality)
         }
+
         if (!launchedAnyTask) {
-          taskSet.abortIfCompletelyBlacklisted(hostToExecutors)
-        }
+          taskSet.getCompletelyBlacklistedTaskIfAny(hostToExecutors) match {
+            case taskIndex: Some[Int] => // Returns the taskIndex which was unschedulable
+
+              // If the taskSet is unschedulable we kill an existing blacklisted executor/s and
+              // kick off an abortTimer which after waiting will abort the taskSet if we were
+              // unable to schedule any task from the taskSet.
+              // Note: We keep a track of schedulability on a per taskSet basis rather than on a
+              // per task basis.
+              val executor = hostToExecutors.valuesIterator.next().iterator.next()
+              logDebug("Killing executor because of task unschedulability: " + executor)
+              blacklistTrackerOpt.foreach(blt => blt.killBlacklistedExecutor(executor))
+
+              if (!unschedulableTaskSetToExpiryTime.contains(taskSet)) {
+                  unschedulableTaskSetToExpiryTime(taskSet) = clock.getTimeMillis()
+                  abortTimer.schedule(new TimerTask() {
+                    override def run() {
+                      if (unschedulableTaskSetToExpiryTime.contains(taskSet) &&
+                        (unschedulableTaskSetToExpiryTime(taskSet)
+                          + UNSCHEDULABLE_TASKSET_TIMEOUT_MS)
+                          <= clock.getTimeMillis()
+                      ) {
+                        logInfo("Cannot schedule any task because of complete blacklisting. " +
+                          "Wait time for scheduling expired. Aborting the application.")
+                        taskSet.abortSinceCompletelyBlacklisted(taskIndex.get)
+                      } else {
+                        this.cancel()
+                      }
+                    }
+                  }, UNSCHEDULABLE_TASKSET_TIMEOUT_MS)
+                }
+            case _ => // Do nothing.
+            }
+          } else {
+            // If a task was scheduled, we clear the expiry time for the taskSet. The abort timer
+            // checks this entry to decide if we want to abort the taskSet.
+            if (unschedulableTaskSetToExpiryTime.contains(taskSet)) {
```

Comment:
you can move this up to the `else` so its an `else if`.  Or you could also just call `remove` without checking `contains`, that avoids probing twice.